index.ts

```diff
@@ -104,6 +115,45 @@ class SHACLValidator {
   validateNodeAgainstShape(focusNode: Term, shapeNode: Term) {
     return this.nodeConformsToShape(focusNode, shapeNode, this.validationEngine)
   }
+
+  private async loadOwlImports() {
+    if (this.importsLoaded) {
+      return
+    }
+
+    this.importsLoaded = true
+
+    const { owl } = this.ns
+    const loaded: Set<NamedNode> = new TermSet()
+
+    const doLoad = async (url: NamedNode) => {
+      if (!this.importGraph) {
+        throw new Error('importGraph parameter is required to load owl:imports')
+      }
+
+      const imported = await this.importGraph(url)
+
+      for (const quad of imported) {
+        this.$shapes.dataset.add(quad)
+      }
+
+      return imported
+    }
+
+    const loadFromDataset = async (dataset: DatasetCore) => {
+      await Promise.all([...dataset
+        .match(null, owl.imports)]
```

Comment:
loading the imports in parallel is probably a good idea indeed. The downside is that in case a dataset contains two imports for the same URL, it will be loaded twice. But:

- this is highly unlikely to happen
- loading twice would be useless but not wrong (idempotency is ensured by set semantics)

Reply:
> case a dataset contains two imports for the same URL,

Hm, let me add a test case like that. I think it should work correctly, because I think that loading same URL twice would potentially be harmful

Reply:
what a about adding a `distinct` to ensure we're not loading the same URL twice?

Reply:
I moved the code which checks what has already been loaded outside the promise. This is probably not perfect but should be a little more accurate